lib/src/components/AbstractDialog/AbstractDialog.tsx

```diff
@@ -0,0 +1,213 @@
+    hasNativeModalSupport = !!(hasWindow && dynamicallyTypedWindow.HTMLDialogElement &&
+      dynamicallyTypedWindow.HTMLDialogElement.prototype.showModal);
+
+const createCancelEvent = () => new Event('cancel', { cancelable: true });
+
+/**
+ * Abstracted Dialog element implementation and behaviors.
+ * @param className - A classname string for the dialog element.
+ * @param onCancel - A callback function that gets called when the dialog is canceled.
+ * @param cancelOnClickOutside - A boolean that if true triggers onCancel when the user clicks outside the dialog.
+ * @param cancelOnClickOutsideTargetClassName - A classname, if specified, selects an element inside the dialog
+ *    element that becomes the reference element for when the user clicks outside of, triggers the cancelOnClickOutside.
+ * @param useNativeCancelOnEscape - If this is set to true, it will attempt to use native dialog element
+ *    cancel behavior when escape is pressed. By default this is set to false, which means it uses event.preventDefault
+ *    and calls onCancel callback when escape is pressed.
+ * @return - Abstracted dialog element.
+ */
+
+// propTypes static analysis doesn't work with the way this component is written
+/* eslint-disable react/prop-types */
+const AbstractDialog = forwardRef<HTMLDialogElement, Props>((props, ref) => {
+  const {
+    className,
+    children,
+    cancelOnClickOutside,
+    onCancel,
+    useNativeCancelOnEscape,
+    role,
+    cancelOnClickOutsideTargetClassName,
+    ...attrs
+  } = props;
+
+  const dialogRef = useRef<HTMLDialogElement>(null);
+
+  // The dialogRef value needs to get passed down in a context. But the context needs to know when the ref
+  // value has updated, and refs aren't tracked like state values. So we have to copy the ref value into a state
+  // value in order for it to be tracked.
+  const [dialogRefState, setDialogRefState] = useState<HTMLDialogElement | null>(null);
+
+  useImperativeHandle(ref, () => dialogRef.current as HTMLDialogElement);
+
+  function dialogKeydownListener(evt: KeyboardEvent<HTMLDialogElement>) {
+    if (evt.key === 'Escape' || evt.key === 'Esc') {
+      // HACK for backwards compatibility: it is known that some downstream uses of NxModal do not provide
+      // the onCancel/onCancel handler despite it being required. It is also known that these downstream uses
+      // have some logic which globally manages ESC handlers for a variety of things including modals. To keep that
+      // working, only stopPropagation if the onCancel callback is defined. If it isn't defined, tenuously assume
+      // that the ESC handling is implemented externall and do nothing here.
+      if (onCancel) {
+        // prevent visibility of the keydown outside of the dialog, so that global ESC listeners on the
+        // document don't pick it up
+        evt.stopPropagation();
+        evt.nativeEvent.stopImmediatePropagation();
+
+        if (evt.defaultPrevented) {
+          return;
```

Comment:
I'm not a fan of early returns.  Would you be willing to rewrite it so we're checking `!evt.defaultPrevented` and all of the remaining code is within that `if` block?

Reply:
> I'm not a fan of early returns. Would you be willing to rewrite it so we're checking `!evt.defaultPrevented` and all of the remaining code is within that `if` block?

Fixed here: f4b19a5fd1c62a95d272a0fd18649c96303ca684